main.cpp

```diff
@@ -21,38 +20,136 @@ int main(int argc, char **argv)
 
         // check validity
         if (filter == '?') {
-            throw std::string {"Invalid filter."};
+            throw string {"Invalid filter."};
         }
 
         // Ensure only one filter
         if (getopt(argc, argv, filtersOptions) != -1) {
-            throw std::string {"Only one filter allowed."};
+            throw string {"Only one filter allowed."};
         }
 
         // Ensure proper usage
         if (argc != optind + 2) {
-            throw std::string {"Usage: filter [flag] infile outfile"};
+            throw string {"Usage: filter [flag] infile outfile"};
         }
 
-        // Open input file and check
-        FILE *inptr {fopen(argv[optind], "r")};
-        if (inptr == nullptr) {
-            throw std::string {"Could not read : " + static_cast<std::string>(argv[optind])};
+        // Open input file 
+        FILE *inFile {fopen(argv[optind], "r")};
+        // check readability
+        if (inFile == nullptr) {
+            throw string {"Could not read : " + static_cast<string>(argv[optind])};
         }
 
         // open output file
-        FILE *outptr {fopen(argv[optind + 1], "w")};
-        if (outptr == nullptr) {
-            throw std::string {"Could not create : " + static_cast<std::string>(argv[optind + 1])};
+        FILE *outFile {fopen(argv[optind + 1], "w")};
+        // check writability
+        if (outFile == nullptr) {
+            throw string {"Could not create : " + static_cast<string>(argv[optind + 1])};
         }
 
+        // Read inFile's BITMAPFILEHEADER
+        BITMAPFILEHEADER bf;
+        fread(&bf, sizeof(BITMAPFILEHEADER), 1, inFile);
+
+        // Read inFile's BITMAPINFOHEADER
+        BITMAPINFOHEADER bi;
+        fread(&bi, sizeof(BITMAPINFOHEADER), 1, inFile);
+
+        // Ensure infile is (likely) a 24-bit uncompressed BMP 4.0
+        if (bf.bfType != 0x4d42 || bf.bfoffBits != 54 || bi.biSize != 40 || bi.biBitCount != 24 || bi.biCompression != 0) {
+            fclose(outFile);
+            fclose(inFile);
+            
+            throw string {"Unsupported file format."};
+        }
+
+        int height {abs(bi.biHeight)};
+        int width {abs(bi.biWidth)};
+
+        // Allocate memory for image
+        RGBTRIPLE *image {new RGBTRIPLE[height * width]};
+        
+        // checks for successful memory allocation
+        if (image == nullptr) {
+            fclose(inFile);
+            fclose(outFile);
+
+            throw string {"Not enough memory to store image."};
+        }
+
+        // Determine padding for scanlines
+        size_t padding {(4 - (width * sizeof(RGBTRIPLE)) % 4) % 4};
```

Comment:
@chandrakant100 @Ank1taS Needs an explanation of this line.

Reply:
Here padding is used for scanline
BMP thus takes up (3 pixels) × (3 bytes per pixel) = 9 bytes, which is not a multiple of 4. And so the scanline is "padded" with as many zeroes as it takes to extend the scanline's length to a multiple of 4. In other words, between 0 and 3 bytes of padding are needed for each scanline in a 24-bit BMP. @Himanshu40 @chandrakant100 you guys better open a separate discussion regarding `padding`

However, padding is always an unsigned integer value, and declaring the padding variable as an **unsigned int** results narrowing conversion warning as follows

```
warning: narrowing conversion of ‘((4 - ((((long unsigned int)width) * sizeof (RGBTRIPLE)) % 4)) % 4)’ from ‘long unsigned int’ to ‘unsigned int’ inside { } [-Wnarrowing]
         unsigned int padding {(4 - (width * sizeof(RGBTRIPLE)) % 4) % 4};
``` 

That's why I declared the padding variable as `size_t` and main.cpp compiled fine.